ktor-server/ktor-server-plugins/ktor-server-cors-common/jvmAndNix/src/io/ktor/server/plugins/cors/CORSCommon.kt

```diff
@@ -0,0 +1,561 @@
+            }
+        }
+    }
+
+    private fun addHost(host: String) {
+        validateWildcardRequirements(host)
+        hosts.add(host)
+    }
+
+    private fun validateWildcardRequirements(host: String) {
+        if ('*' !in host) {
+            return
+        }
+
+        fun String.countMatches(subString: String): Int =
+            windowed(subString.length) {
+                if (it == subString) 1 else 0
+            }.sum()
+
+        require(wildcardInFrontOfDomain(host)) { "wildcard must appear in front of the domain, e.g. *.domain.com" }
+        require(host.countMatches(wildcardWithDot) == 1) { "wildcard cannot appear more than once" }
+    }
+
+    private fun wildcardInFrontOfDomain(host: String): Boolean {
+        val indexOfWildcard = host.indexOf(wildcardWithDot)
+        return wildcardWithDot in host && !host.endsWith(wildcardWithDot) &&
+            (indexOfWildcard <= 0 || host.substringBefore(wildcardWithDot).endsWith("://"))
+    }
+
+    /**
+     * Allows exposing the [header] using `Access-Control-Expose-Headers`.
+     * The `Access-Control-Expose-Headers` header adds the specified headers
+     * to the allowlist that JavaScript in browsers can access.
+     */
+    public fun exposeHeader(header: String) {
+        if (header !in CorsSimpleResponseHeaders) {
+            exposedHeaders.add(header)
+        }
+    }
+
+    /**
+     * Allows using the `X-Http-Method-Override` header for the actual [CORS] request.
+     */
+    @Suppress("unused")
+    public fun allowXHttpMethodOverride() {
+        allowHeader(HttpHeaders.XHttpMethodOverride)
+    }
+
+    /**
+     * Allows using headers prefixed with [headerPrefix] for the actual [CORS] request.
+     */
+    public fun allowHeadersPrefixed(headerPrefix: String) {
+        val prefix = headerPrefix.lowercase()
+        this.headerPredicates.add { name -> name.startsWith(prefix) }
+    }
+
+    /**
+     * Allows using headers matching [predicate] for the actual [CORS] request.
+     */
+    public fun allowHeaders(predicate: (String) -> Boolean) {
+        this.headerPredicates.add(predicate)
+    }
+
+    /**
+     * Allow using a specified [header] for the actual [CORS] request.
+     */
+    public fun allowHeader(header: String) {
+        if (header.equals(HttpHeaders.ContentType, ignoreCase = true)) {
+            allowNonSimpleContentTypes = true
+            return
+        }
+
+        if (header !in CorsSimpleRequestHeaders) {
+            headers.add(header)
+        }
+    }
+
+    /**
+     * Adds a specified [method] to a list of methods allowed by [CORS].
+     *
+     * Note that CORS operates with real HTTP methods only and
+     * doesn't handle method overridden by `X-Http-Method-Override`.
+     */
+    public fun allowMethod(method: HttpMethod) {
+        if (method !in CorsDefaultMethods) {
+            methods.add(method)
+        }
+    }
+}
+
+internal enum class OriginCheckResult {
```

Comment:
consider extracting internal API to a separate file. It can make code easier to read